hazelcast/src/main/java/com/hazelcast/sql/impl/expression/math/RandFunction.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.expression.math;
+
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+import com.hazelcast.sql.impl.SqlDataSerializerHook;
+import com.hazelcast.sql.impl.expression.Expression;
+import com.hazelcast.sql.impl.expression.ExpressionEvalContext;
+import com.hazelcast.sql.impl.expression.UniExpression;
+import com.hazelcast.sql.impl.row.Row;
+import com.hazelcast.sql.impl.type.QueryDataType;
+
+import java.util.Random;
+import java.util.concurrent.ThreadLocalRandom;
+
+/**
+ * Random function implementation.
+ */
+public class RandFunction extends UniExpression<Double> implements IdentifiedDataSerializable {
+    public RandFunction() {
+        // No-op.
+    }
+
+    public RandFunction(Expression<?> seedExp) {
+        super(seedExp);
+    }
+
+    public static RandFunction create(Expression<?> seedExp) {
+        return new RandFunction(seedExp);
+    }
+
+    @Override
+    public Double eval(Row row, ExpressionEvalContext context) {
+        Long seed = operand != null ? MathFunctionUtils.asBigint(operand, row, context) : null;
+
+        Random random = seed != null ? new Random(seed) : ThreadLocalRandom.current();
```

Comment:
It might make sense to create `Random` once if `seed` is provided and use `setSeed` in `eval` instead of allocating a new `Random` on every invocation.

What it the semantics of `seed`? Calcite's `SqlRandFunction` and `RandomFunction` define it as "initializing the random number generator with seed on first call".

BTW, what thread safety level is expected for expressions? As far as I remember, `Exec` is not required to be thread safe and expressions belong to its state.

Reply:
Naive question: why `Random` instead of `SecureRandom`?

Reply:
@taburet `seed` could change from invocation to invocation. We do not know whether it should be used or not before the actual operand is processed. I would not complicate things, this is a micro-optimization for a function, which is mainly used for testing.

Reply:
@nfrankel This function doesn't have a requirement to be cryptographically strong, therefore we use the more performant implementation.

Reply:
@devozerov what about semantics of `seed` and implied thread safety for expressions?

Reply:
Expressions do not have a shared state - does this answer the question?

Reply:
Does that mean there should be no state to share or the state is not shared between threads?

Reply:
One expression instance is not meant to have any semantical influence on the results of other expressions.

Reply:
To put it simple, can `eval` be invoked by the query execution engine from more than one thread simultaneously on the same expression instance?

Reply:
No. Expressions are not thread-safe. A single instance is not meant to be executed by multiple threads simultaneously.